test/test_distributed.py

```diff
@@ -700,13 +747,89 @@ def test_all_gather_multigpu(self):
+        model.train()
+        output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_multigpu
+    def test_DistributedDataParallel(self):
+        # Run a simple end to end DDP model, use result of single node model
+        # as baseline
+        group, group_id, rank = self._init_global_test()
+        rankToGPUMapping = self._init_multigpu_helper()
+
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.fc1 = nn.Linear(2, 10, bias=False)
+                self.fc2 = nn.Linear(10, 50, bias=False)
+                self.fc3 = nn.Linear(50, 4, bias=False)
+                self.relu = nn.ReLU()
+
+            def forward(self, x):
+                x = self.relu(self.fc1(x))
+                x = self.relu(self.fc2(x))
+                x = self.fc3(x)
+                return F.softmax(x, dim=1)
+
+        # cpu training setup
+        model = Net()
+
+        # single gpu training setup
+        model_gpu = copy.deepcopy(model)
+        gpu_subset = list(rankToGPUMapping[rank])
+        model_gpu.cuda(gpu_subset[0])
+
+        # DDP training setup
+        model_DDP = copy.deepcopy(model)
+        model_DDP.cuda(gpu_subset[0])
+        model_DDP = nn.parallel.DistributedDataParallel(model_DDP, device_ids=gpu_subset)
+        optimizer = optim.SGD(model_DDP.parameters(), lr=1e-2)
+        optimizer.zero_grad()
+
+        # batch_size for DDP should be divisible by #GPU per node.
+        batch_size = len(gpu_subset) * int(WORLD_SIZE)
+        input_cpu = torch.randn(batch_size, 2)
+        target = torch.randn(batch_size, 4)
+        loss = nn.MSELoss()
+
+        # single gpu training
+        self._test_DDP_helper(model_gpu,
+                              input_cpu.cuda(gpu_subset[0]),
+                              target.cuda(gpu_subset[0]),
+                              loss)
+
+        # DDP training, DDP scatters subsets of input_cpu to nodes/GPUs
+        self._test_DDP_helper(model_DDP,
+                              input_cpu[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)],
+                              target[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)].cuda(gpu_subset[0]),
+                              loss)
+
+        for layer_gpu, layer_DDP in zip(model_gpu.modules(), model_DDP.module.modules()):
+            if isinstance(layer_gpu, nn.Linear):
+                self.assertEqual(layer_gpu.weight.grad, layer_DDP.weight.grad)
+
+        # Run SGD and second iteration to shake out errors
+        optimizer.step()
+        self._test_DDP_helper(model_DDP,
+                              input_cpu,
+                              target.cuda(gpu_subset[0]),
+                              loss)
+
```

Comment:
We're not checking that the grads match after the second iteration 😕 That's important